src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -377,31 +377,53 @@ public static bool IsPointerOrFunctionPointer(this TypeSymbol type)
             return type.IsDelegateType() ? (NamedTypeSymbol)type : null;
         }
 
-        public static TypeSymbol? GetDelegateOrFunctionPointerType(this TypeSymbol type)
+        public static TypeSymbol? GetDelegateOrFunctionPointerType(this TypeSymbol? type)
         {
             return (TypeSymbol?)GetDelegateType(type) ?? type as FunctionPointerTypeSymbol;
         }
 
         /// <summary>
         /// return true if the type is constructed from System.Linq.Expressions.Expression`1
         /// </summary>
-        public static bool IsExpressionTree(this TypeSymbol _type)
+        public static bool IsExpressionTree(this TypeSymbol type)
         {
-            return _type.OriginalDefinition is NamedTypeSymbol type &&
-                type.Arity == 1 &&
-                type.MangleName &&
-                type.Name == "Expression" &&
-                CheckFullName(type.ContainingSymbol, s_expressionsNamespaceName);
+            return type.IsGenericOrNonGenericExpressionType(out bool isGenericType) && isGenericType;
+        }
+
+        public static bool IsNonGenericExpressionType(this TypeSymbol type)
+        {
+            return type.IsGenericOrNonGenericExpressionType(out bool isGenericType) && !isGenericType;
         }
 
+        public static bool IsGenericOrNonGenericExpressionType(this TypeSymbol _type, out bool isGenericType)
+        {
+            if (_type.OriginalDefinition is NamedTypeSymbol type &&
+                type.Name == "Expression" &&
+                CheckFullName(type.ContainingSymbol, s_expressionsNamespaceName))
+            {
+                if (type.Arity == 0)
+                {
+                    isGenericType = false;
+                    return true;
+                }
+                if (type.Arity == 1 &&
+                    type.MangleName)
```

Comment:
nit: this is existing logic, but isn't it redundant with the `Arity` check?

Reply:
> this is existing logic, but isn't it redundant with the Arity check?

This check will ensure we don't match to a generic type named `Expression` (without the trailing ````1```), although that scenario is unlikely. Since this is an existing check, I'll leave it as is.